src/Lykke.Service.PayInternal.Services/BtcTransferRequestService.cs

```diff
@@ -0,0 +1,428 @@
+using Lykke.Service.PayInternal.Core.Domain.Wallet;
+using Lykke.Service.PayInternal.Core.Services;
+using Lykke.Service.PayInternal.Services.Domain;
+using NBitcoin;
+using Newtonsoft.Json;
+
+
+namespace Lykke.Service.PayInternal.Services
+{
+    public class BtcTransferRequestService : ITransferRequestService
+    {
+
+        private readonly ITransferRepository _transferRepository;
+        private readonly IWalletRepository _walletRepository;
+        private readonly IBitcoinApiClient _bitcointApiClient;
+        private readonly ITransferRequestPublisher _transferRequestPublisher;
+        private readonly ILog _log;
+
+
+        #region .ctors
+        public BtcTransferRequestService(
+            ITransferRepository transferRepository,
+            IWalletRepository walletRepository,
+            IBitcoinApiClient bitcointApiClient,
+            ITransferRequestPublisher transferRequestPublisher,
+            ILog log)
+        {
+            _transferRepository = transferRepository ?? throw new ArgumentNullException(nameof(transferRepository));
+            _walletRepository = walletRepository ?? throw new ArgumentNullException(nameof(walletRepository));
+            _bitcointApiClient = bitcointApiClient ?? throw new ArgumentNullException(nameof(bitcointApiClient));
+            _transferRequestPublisher = transferRequestPublisher ?? throw new ArgumentNullException(nameof(transferRequestPublisher));
+            _log = log ?? throw new ArgumentNullException(nameof(log));
+        }
+        #endregion
+
+        #region public
+
+        /// <summary>
+        /// Tries to execute a crosswise transfer request from many sources to many destinations.
+        /// </summary>
+        /// <param name="transferRequest">The transfer request.</param>
+        /// <returns>The updated transfer request OR transfer request with error.</returns>
+        public async Task<ITransferRequest> CreateTransferCrosswiseAsync(ITransferRequest transferRequest)
+        {
+            // We need an independent copy of the request here, for some methods in current implementation change the inner state of the given object.
+            // For instance, the BtcTransferRequest class' constructor reuses the transaction list it gets.
+            var transferRequestUpdated = transferRequest.DeepCopy();
+
+            // First of all, we need to verify the source addresses to belong to the Merchant.
+            var sourcesCheckupResult = await CheckupSourcesAndWalletsAsync(transferRequestUpdated);
+            if (sourcesCheckupResult != null)
+                return sourcesCheckupResult;
+
+            // Here would be no additional validations. If transaction fails, we will know it from the underlying BitcoinApi service.
+
+            // Save the initial state of transfer
+            await _transferRepository.SaveAsync(transferRequestUpdated);
+            await _transferRequestPublisher.PublishAsync(transferRequestUpdated);
```

Comment:
No need to publish changes 3 times per method. Publish'em once you finished the entity updates.